Fix turrets aiming through walls

joveeater opened the pull request:
#### Summary

SUMMARY: Bugfixes "Fix turrets aiming through walls"

#### Purpose of change

Fixes #2578 fixes #2431

#### Describe the solution

Bit of an overhaul to their targeting to make it a little better at selecting the closest target and a bit clearer. 

#### Describe alternatives you've considered

Just adding the missing vision check.

#### Testing

I had a drive around some turrets and they're firing correctly now.

scarf005 reviewed (approved):
![image](https://user-images.githubusercontent.com/54838975/230702375-c5d08341-7038-4341-b474-53d68042c377.png)
confirmed that turrets do not shoot at vehicles when blocked by walls. (casings are from testing prev. behavior)

wish std::optional were used, but it'd be better to just use ranges-v3 at that point.

Firestorm01X2 commented:
@joveeater For me turret now do not even shoot vehicle. Tested on APC. Save attached:
[turret.zip](https://github.com/cataclysmbnteam/Cataclysm-BN/files/11186164/turret.zip)

metalheimer commented:
>  (casings are from testing prev. behavior)

It's not fixed for me. Tried a very recent version (no game mods either). Are you sure those bunker turrets weren't just out of ammo? Bunker turrets still try to pepper me through walls. Tried with a car, and a bicycle. They shoot every time.

scarf005 commented:
hecc. recompile time.

scarf005 commented:
![image](https://user-images.githubusercontent.com/54838975/230907847-ca70cdb9-c129-4036-abbe-2bc60043a7bc.png)
@metalheimer that's because this pull request is not yet merged.

scarf005 commented on src/mattack_actors.cpp:
```diff
@@ -481,21 +481,42 @@ int gun_actor::get_max_range()  const
+        bool found_controls = false;
+
+        for( const vpart_reference &vp : v.v->get_avail_parts( "CONTROLS" ) ) {
+            if( z.sees( vp.pos() ) ) {
```
![image](https://user-images.githubusercontent.com/54838975/230907989-a0498039-8e2b-4aec-a13c-e96d933579a3.png)
![image](https://user-images.githubusercontent.com/54838975/230908288-e855d7d7-06d6-4cf1-b080-018034c550c6.png)

This line does not account for vehicle parts that obstruct views.
- bicycle test worked because its control was visible from outside.
- @Firestorm01X2's APC test did not work because quarterpanels were blocking the vehicle control from the view of the turret.

joveeater commented:
I've fixed the APC problem. Good catch. It no longer lets a vehicle block the view to that same vehicle. I also refactored it to use optional, was a good suggestion.

scarf005 reviewed (approved):
|       bicycle       |     APC     |
| :-----------------: | :---------: |
| ![bicycle][bicycle] | ![APC][apc] |

|   turret    | bicycle(stopped) | bicycle(moving) | APC(stopped) | APC(moving) |
| :---------: | :--------------: | :-------------: | :----------: | :---------: |
|   normal    |      shoots      |     shoots      |              |   shoots    |
| behind wall |                  |                 |              |             |

looks like it's working as intended.

[apc]: https://user-images.githubusercontent.com/54838975/230912935-fabcdd47-866a-45bc-a36a-6a7395fbbb9e.png
[bicycle]: https://user-images.githubusercontent.com/54838975/230913818-d47e4c24-cd2f-4631-bd4d-cdd2a5b2ff10.png

joveeater commented:
Urgh, this still doesn't handle floors correctly atm. I'll take another look.

Firestorm01X2 commented:
> looks like it's working as intended.

I'll double check. No idea why it does no for my save. I've used light turret and apc to test.

Firestorm01X2 commented:
> Urgh, this still doesn't handle floors correctly atm. I'll take another look.

I'll convert it to draft to prevent early merging.

scarf005 commented:
i think using `draft` would be better since it's official github feature.
